Update documentation on adding spatial masks in user-defined modules (#1567)
nest/nest-simulator: doc/guides/spatial/guide_spatially_structured_networks.rst

    :start-after: #{ vislayer #}
    :end-before: #{ end #}

.. _ch:custom_masks:

Creating custom masks
---------------------

> In some cases, the built-in masks may not meet your needs, and you want to
> create a custom mask. There are two ways to do this: the first is to use Parameters
> to introduce a cut-off to the connection probability. The second is to implement a custom
> mask in C++ as a module.

Reviewer: ```suggestion
In some cases, the built-in masks may not meet your needs, and you want to
create a custom mask. There are two ways to do this: 

1. To use Parameters to introduce a cut-off to the connection probability. 
2. To implement a custom mask in C++ as a module.
```

> Using Parameters is the most accessible option; the entire implementation is done on

Reviewer: Could we write ``parameters`` in lowercase throughout the guide? That's the letter case we have also used in the "NEST 3.0: What's new?" guide.

Author: Done.

the PyNEST level. However, the price for this flexibility is reduced connection efficiency
compared to masks implemented in C++. Combining Parameters to give the wanted behaviour may
also be difficult if the mask specifications are complex.

Implementing a custom mask in C++ gives much higher connection performance and greater freedom
in implementation, but requires some knowledge of the C++ language. As the mask in this case
is implemented in an extension module, which is dynamically loaded into NEST, it also requires
some additional steps for installation.

.. _sec:maskparameter:

Using Parameters to specify connection boundaries
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can use Parameters that represent spatial distances between nodes to create a connection

> probability with mask behaviour. For this you need to create a condition Parameter that describes

Reviewer: ```suggestion
probability with mask behaviour. For this, you need to create a condition Parameter that describes
```

the boundary of the mask. As condition Parameters evaluate to either ``0`` or ``1``, it can be

> used alone, used with the ``nest.logic.conditional()`` Parameter, or multiplied with another

Reviewer: ```suggestion
used alone, with the ``nest.logic.conditional()`` Parameter, or multiplied with another
```

Parameter or value, before passing it as the connection probability.

> As an example, we will define an elliptical connection boundary using Parameters. First we define

Reviewer: ```suggestion
As an example, we will define an elliptical connection boundary using Parameters. First, we define
```

variables controlling the shape of the ellipse.

::

   rx = 0.5   # radius in the x-direction
   ry = 0.25  # radius in the y-direction

> Next we define the connection boundary. We only want to connect to targets inside an ellipse, so

Reviewer: ```suggestion
Next, we define the connection boundary. We only want to connect to targets inside an ellipse, so
```

the condition is

.. math::

   \frac{x^2}{r_x^2}+\frac{y^2}{r_y^2} \leq 1,

where :math:`x` and :math:`y` are the distances between the source and target neuron, in x- and
y-directions, respectively. We use this expression to define the boundary using Parameters.

::

   x = nest.spatial.distance.x
   y = nest.spatial.distance.y
   lhs = x * x / rx**2 + y * y / ry**2
   mask_param = nest.logic.conditional(lhs <= 1.0, 0.5, 0.0)
   # Could also have defined it as
   # mask_param = 0.5*(lhs <= 1.0)

> Then we can use the Parameter as connection probability when connecting populations with spatial

Reviewer: ```suggestion
Then, we can use the Parameter as connection probability when connecting populations with spatial
```

You can use parameters that represent spatial distances between nodes to create a connection
probability with mask behaviour. For this, you need to create a condition parameter that describes
the boundary of the mask. As condition parameters evaluate to either ``0`` or ``1``, it can be
used alone, with the ``nest.logic.conditional()`` parameter, or multiplied with another
parameter or value, before passing it as the connection probability.

As an example, we will define an elliptical connection boundary using parameters. First, we define
variables controlling the shape of the ellipse.

::

   rx = 0.5   # radius in the x-direction
   ry = 0.25  # radius in the y-direction

Next, we define the connection boundary. We only want to connect to targets inside an ellipse, so
the condition is

.. math::

   \frac{x^2}{r_x^2}+\frac{y^2}{r_y^2} \leq 1,

where :math:`x` and :math:`y` are the distances between the source and target neuron, in x- and
y-directions, respectively. We use this expression to define the boundary using parameters.

::

   x = nest.spatial.distance.x
   y = nest.spatial.distance.y
   lhs = x * x / rx**2 + y * y / ry**2

>    mask_param = nest.logic.conditional(lhs <= 1.0, 0.5, 0.0)

Reviewer 2: I think you should have a sentence further up that if `x` or `y` falls inside the boundary, we still only connect with a 50% probablility. So that people understand where `0.5` and `0.0` comes in. :)

Author: Done!

   # Could also have defined it as

>    # mask_param = 0.5*(lhs <= 1.0)

Reviewer 2: So you don't need to use `nest.logic.conditional()` at all when sending down to NEST? It is converted? This we should document elsewhere as well. Not in this PR though!

Author: In cases where the resulting value of `nest.logic.conditional()` is `0` if the condition evaluates to false, it is possible to just use multiplication with the condition instead. That's because the condition also is a parameter, which actually evaluates to `0` or `1`.

I suppose we should have some documentation on parameter comparators somewhere, probably in the detailed transition guide.

<http://nest.github.io/nest-simulator/extension_modules>`__ in the NEST
Developer Manual. The basic steps required to get started are:

> 1. From the NEST source directory, copy the directory examples/MyModule to

Reviewer: ```suggestion
1. From the NEST source directory, copy the directory ``examples/MyModule`` to
```

   somewhere outside the NEST source, build or install directories.

   .. code:: bash

      cp -r <path_to_nest_source>/examples/MyModule mm_src

> 2. Create a build directory for MyModule, e.g. mm_bld, next to

Reviewer: ```suggestion
2. Create a build directory for MyModule, e.g. ``mm_bld``, next to
```

   the copied source directory:

   .. code:: bash

      mkdir mm_bld
      cd mm_bld

3. Configure. The configure process uses the script ``nest-config`` to
   find out where NEST is installed, where the source code resides, and
   which compiler options were used for compiling NEST. If
   ``nest-config`` is not in your path, you need to provide it

>    explicitly with ``-Dwith-nest`` like this

Reviewer: ```suggestion
   explicitly with ``-Dwith-nest`` as follows:
```

   .. code:: bash

      cmake -Dwith-nest=${NEST_INSTALL_DIR}/bin/nest-config ../mm_src

> 4. MyModule will then be installed to ``\${NEST_INSTALL_DIR}``. This

Author: Yep, fixed.
